src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Hover/RazorHoverEndpoint.cs

```diff
@@ -5,88 +5,95 @@
         private readonly RazorHoverInfoService _hoverInfoService;
+        private readonly RazorDocumentMappingService _documentMappingService;
         private VSInternalClientCapabilities? _clientCapabilities;
 
         public RazorHoverEndpoint(
             DocumentContextFactory documentContextFactory,
             RazorHoverInfoService hoverInfoService,
+            LanguageServerFeatureOptions languageServerFeatureOptions,
+            RazorDocumentMappingService documentMappingService,
+            ClientNotifierServiceBase languageServer,
             ILoggerFactory loggerFactory)
+            : base(documentContextFactory, languageServerFeatureOptions, documentMappingService, languageServer, loggerFactory)
         {
-            if (documentContextFactory is null)
-            {
-                throw new ArgumentNullException(nameof(documentContextFactory));
-            }
+            _hoverInfoService = hoverInfoService ?? throw new ArgumentNullException(nameof(hoverInfoService));
+            _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
+        }
 
-            if (hoverInfoService is null)
-            {
-                throw new ArgumentNullException(nameof(hoverInfoService));
-            }
+        public RegistrationExtensionResult? GetRegistration(VSInternalClientCapabilities clientCapabilities)
+        {
+            const string AssociatedServerCapability = "hoverProvider";
+            _clientCapabilities = clientCapabilities;
 
-            if (loggerFactory is null)
+            var registrationOptions = new HoverOptions()
             {
-                throw new ArgumentNullException(nameof(loggerFactory));
-            }
+                WorkDoneProgress = false,
+            };
 
-            _documentContextFactory = documentContextFactory;
-            _hoverInfoService = hoverInfoService;
-            _logger = loggerFactory.CreateLogger<RazorHoverEndpoint>();
+            return new RegistrationExtensionResult(AssociatedServerCapability, registrationOptions);
         }
 
-        public async Task<VSInternalHover?> Handle(VSHoverParamsBridge request, CancellationToken cancellationToken)
+        /// <inheritdoc/>
+        protected override string EndpointName => LanguageServerConstants.RazorHoverEndpointName;
+
+        /// <inheritdoc/>
+        protected override DelegatedHoverParams CreateDelegatedParams(VSHoverParamsBridge request, DocumentContext documentContext, Projection projection)
+            => new DelegatedHoverParams(
+                documentContext.Identifier,
+                projection.Position,
+                projection.LanguageKind);
+
+        /// <inheritdoc/>
+        protected override Task<VSInternalHover?> HandleInRazorAsync(VSHoverParamsBridge request, DocumentContext documentContext, RazorCodeDocument codeDocument, SourceText sourceText, CancellationToken cancellationToken)
```

Comment:
I can't recall if we have custom logic for `InRazor` scenarios. IIRC I think we always bail. Only bringing it up because I could imagine `CreateDelegatedParams` be `TryCreateDelegatedParams` which would return `false` in the case of Razor allowing us to short circuit the delegation pass

Reply:
For hover we do, and I think for rename we will now too (I might have broken that). Check next iteration and see what you think